src/plugins/intel_gpu/src/graph/graph_optimizer/select_preferred_formats.cpp

```diff
@@ -0,0 +1,77 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+
+#include "pass_manager.h"
+#include "data_inst.h"
+#include "mutable_data_inst.h"
+#include "program_node.h"
+#include "intel_gpu/runtime/engine.hpp"
+#include "runtime/cldnn_itt.hpp"
+#include <iostream>
+#include "to_string_utils.h"
+
+#ifdef ENABLE_ONEDNN_FOR_GPU
+#include <oneapi/dnnl/dnnl.hpp>
+#include "intel_gpu/runtime/debug_configuration.hpp"
+#include "impls/onednn/utils.hpp"
+#include "impls/onednn/convolution_onednn.hpp"
+#endif
+#include "intel_gpu/runtime/engine.hpp"
+
+
+using namespace cldnn;
+
+void select_preferred_formats::run(program& p) {
+    OV_ITT_SCOPED_TASK(itt::domains::CLDNN, "pass::select_preferred_formats");
+
+    auto& engine = p.get_engine();
+    const auto& device_info = engine.get_device_info();
+
+    if (!device_info.supports_immad)
+        return;
+
+#ifdef ENABLE_ONEDNN_FOR_GPU
+    GPU_DEBUG_GET_INSTANCE(debug_config);
+    for (auto n : p.get_processing_order()) {
+        if (n->is_input() || !layout_optimizer::are_data_types_suitable_for_onednn(*n)) {
+            continue;
+        }
+
+        if (n->is_type<convolution>()) {
+            // Currently, only care for onednn convolutions
+            // Onednn primitive descriptor creation may fail, for example, due to asymmetric weight.
+            try {
+                auto& node = n->as<convolution>();
+                n->init_preferred_fmt(node.get_dependencies().size(), node.get_users().size());
+                for (size_t idx = 0 ; idx < node.get_dependencies().size() ; idx++) {
+                    if (node.get_dependency(idx).is_constant())
+                        continue;
+
+                    auto desc = onednn::get_convolution_descriptor(*node.get_kernel_impl_params(), dnnl::memory::format_tag::any);
+                    // Note: did not handle attribute properly. especially for zero-point
+                    dnnl::primitive_desc prim_desc{&desc->data, nullptr, engine.get_onednn_engine(), nullptr};
```

Comment:
Why primitive desc creation happen in a loop? I think it should be outside.

Reply:
Applied